x-pack/auditbeat/module/system/process/gosysinfo_provider.go

```diff
@@ -0,0 +1,461 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package process
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"runtime"
+	"strconv"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/gofrs/uuid/v5"
+
+	"github.com/elastic/beats/v7/auditbeat/datastore"
+	"github.com/elastic/beats/v7/auditbeat/helper/hasher"
+	"github.com/elastic/beats/v7/libbeat/common/capabilities"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/cache"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/module/system"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
+)
+
+const (
+	bucketName              = "auditbeat.process.v1"
+	bucketKeyStateTimestamp = "state_timestamp"
+)
+
+// SysinfoMetricSet collects data about the host.
+type SysInfoMetricSet struct {
+	system.SystemMetricSet
+	MetricSet
+	hasher    *hasher.FileHasher
+	cache     *cache.Cache
+	bucket    datastore.Bucket
+	lastState time.Time
+
+	suppressPermissionWarnings bool
+}
+
+// Process represents information about a process.
+type Process struct {
+	Info         types.ProcessInfo
+	UserInfo     *types.UserInfo
+	User         *user.User
+	Group        *user.Group
+	CapEffective []string
+	CapPermitted []string
+	Hashes       map[hasher.HashType]hasher.Digest
+	Error        error
+}
+
+// Hash creates a hash for Process.
+func (p Process) Hash() uint64 {
+	h := xxhash.New()
+	//nolint:errcheck // always return nil err
+	h.WriteString(strconv.Itoa(p.Info.PID))
+	//nolint:errcheck // always return nil err
+	h.WriteString(p.Info.StartTime.String())
+	return h.Sum64()
+}
+
+func (p Process) toMapStr() mapstr.M {
+	return mapstr.M{
+		// https://github.com/elastic/ecs#-process-fields
+		"name": p.Info.Name,
+		"args": p.Info.Args,
+		"pid":  p.Info.PID,
+		"parent": mapstr.M{
+			"pid": p.Info.PPID,
+		},
+		"working_directory": p.Info.CWD,
+		"executable":        p.Info.Exe,
+		"start":             p.Info.StartTime,
+	}
+}
+
+// New constructs a new MetricSet.
```

Comment:
Comment should be in the form `// NewFromSysInfo ...`

Reply:
fixed